File: pdl/infra/csharp/Io/Pdl/Infrastructure/Data/RowExecutor.cs
using System.Collections.Generic;

namespace Io.Pdl.Infrastructure.Data;

public static class RowExecutor
{
    public static void Create(Row record)
    {
        var store = record.ResolveStore();
        record.SetStore(store);
        var values = record.CollectValues();
        var inserted = store.Insert(record.Table, record.PrimaryKey, values);
        if (inserted.Count > 0)
        {
            record.ApplyValues(inserted);
        }
    }

    public static void Update(Row record)
    {
        var store = record.ResolveStore();
        record.SetStore(store);
        var values = record.CollectValues();
        store.Update(record.Table, record.PrimaryKey, values);
    }

    public static void Delete(Row record)
    {
        var store = record.ResolveStore();
        record.SetStore(store);
        var values = record.CollectValues();
        store.Delete(record.Table, record.PrimaryKey, values);
    }

    public static void Hydrate(Row record, IDictionary<string, object?> values)
    {
        record.ApplyValues(values);
    }
}

File: pdl/infra/java/src/main/java/io/pdl/infrastructure/data/RowExecutor.java
package io.pdl.infrastructure.data;

import io.pdl.infrastructure.data.RowIntrospector.RowMetadata;

import java.util.Map;

public final class RowExecutor {
    private RowExecutor() {
    }

    public static void create(Row record) throws Exception {
        RowMetadata metadata = RowIntrospector.introspect(record.getClass());
        Map<String, Object> values = metadata.extractValues(record);
        DBStore resolved = Row.resolveStore(record.internalStore());
        record.setInternalStore(resolved);
        Map<String, Object> generated = resolved.insert(record.getTable(), record.getPrimaryKey(), values);
        if (generated != null && !generated.isEmpty()) {
            metadata.apply(record, generated);
        }
    }

    public static void update(Row record) throws Exception {
        RowMetadata metadata = RowIntrospector.introspect(record.getClass());
        Map<String, Object> values = metadata.extractValues(record);
        DBStore resolved = Row.resolveStore(record.internalStore());
        record.setInternalStore(resolved);
        resolved.update(record.getTable(), record.getPrimaryKey(), values);
    }

    public static void delete(Row record) throws Exception {
        RowMetadata metadata = RowIntrospector.introspect(record.getClass());
        Map<String, Object> values = metadata.extractValues(record);
        DBStore resolved = Row.resolveStore(record.internalStore());
        record.setInternalStore(resolved);
        resolved.delete(record.getTable(), record.getPrimaryKey(), values);
    }

    public static void hydrate(Row record, Map<String, Object> values) {
        RowMetadata metadata = RowIntrospector.introspect(record.getClass());
        metadata.apply(record, values);
    }
}
